### ISSUE TITLE
Playwright tests not working with GitHub Action? browserType.launch: Executable doesn't exist at /github/home/.cache/ms-playwright/chromium-1076/chrome-linux/chrome

### ISSUE THREAD

**Issue by GeorgeXCV** [artilleryio/artillery#2198]
Version info: 
```
v2.0.0-37
```

Running this command: 
```
run tests/my-test.yml --output ./report.json
```

I expected to see this happen:

The test to run like it does locally.

Instead, this happened:

```
worker error, id: 1 browserType.launch: Executable doesn't exist at /github/home/.cache/ms-playwright/chromium-1076/chrome-linux/chrome
╔═════════════════════════════════════════════════════════════════════════╗
║ Looks like Playwright Test or Playwright was just installed or updated. ║
║ Please run the following command to download new browsers:              ║
║                                                                         ║
║     npx playwright install                                              ║
║                                                                         ║
║ <3 Playwright Team                                                      ║
╚═════════════════════════════════════════════════════════════════════════╝
    at Array.scenario (/home/node/artillery/node_modules/artillery-engine-playwright/index.js:56:38)
    at runScenario (/home/node/artillery/node_modules/@artilleryio/int-core/lib/runner.js:331:32)
    at EventEmitter.<anonymous> (/home/node/artillery/node_modules/@artilleryio/int-core/lib/runner.js:203:25)
    at EventEmitter.emit (/home/node/artillery/node_modules/eventemitter3/index.js:181:35)
    at UniformProcess.<anonymous> (/home/node/artillery/node_modules/@artilleryio/int-core/lib/phases.js:220:[12](https://github.com/nansen-ai/nansen-artillery/actions/runs/6467513570/job/17557698774#step:9:13))
    at /home/node/artillery/node_modules/arrivals/lib/index.js:47:10
    at NanoTimer.setInterval (/home/node/artillery/node_modules/nanotimer/lib/nanotimer.js:2[20](https://github.com//actions/runs/6467513570/job/17557698774#step:9:21):5)
    at Immediate._onImmediate (/home/node/artillery/node_modules/nanotimer/lib/nanotimer.js:203:68) ***
  name: 'Error'
```

Files being used:
Workflow:

```
name: Load Test

on:
    workflow_dispatch:
 
jobs:
  artillery:
    runs-on: ubuntu-latest
 
    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Set up Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18

      - name: Setup bun dependencies
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest
        
      - name: Install dependencies
        run: |
          bun install
          npx playwright install

      - name: Write the serviceAccountKey.json file 📝
        run: |
            echo '${{ secrets.SERVICE_ACCOUNT_KEY }}' > serviceAccountKey.json
 
      - name: Execute load tests
        uses: artilleryio/action-cli@v1
        with:
          command: run tests/my-test.yml --output ./report.json
        
      - name: Upload artifact
        uses: actions/upload-artifact@v3
        if: always()
        with:
          name: artillery-report
          # Reference the generated report in the file system.
          path: ./report.json
```

**Comment by hassy** [comment]
Try changing `npx playwright install` to `npx playwright@1.37.1 install`. Artillery's Playwright integration uses Playwright v1.37.1 ([source](https://github.com/artilleryio/artillery/blob/main/packages/artillery-engine-playwright/package.json#L14)), and browsers in Playwright need to be installed for specific versions.

**Comment by hassy** [comment]
Or better yet: run your tests on AWS Fargate ([docs](https://www.artillery.io/docs/load-testing-at-scale/aws-fargate)). That will let you generate more load (GitHub Actions is not a great place to actually execute the tests from), and you won't need to install browsers explicitly in the workflow.

**Comment by GeorgeXCV** [comment]
> npx playwright@1.37.1 install

Still get the same error.

**Comment by TestItEver** [comment]
> Or better yet: run your tests on AWS Fargate ([docs](https://www.artillery.io/docs/load-testing-at-scale/aws-fargate)). That will let you generate more load (GitHub Actions is not a great place to actually execute the tests from), and you won't need to install browsers explicitly in the workflow.

Hi @hassy! I've just got the same error on AWS Fargate, although a couple of days ago it works correctly. I didn't update or change anything in my project so far.

**Comment by harv-singh** [comment]
➕ was able to reproduce this

**Comment by nlyn** [comment]
I'm also getting this on Fargate

**Comment by freyjadomville** [comment]
We are also able to repro this with our Playwright test using Fargate, on the latest available release (v2.0.0-38).

**Comment by freyjadomville** [comment]
I *think* I have a way to get this working inside the flow itself so that the initialisation is done once per worker on Fargate - it behaves locally, so I'm hoping that I can execute the same command set on Fargate, depending on if npx itself is executable inside the Fargate cluster instance. I have some local issues to fix in my current Playwright test, but I'll report back separately once I have something working or not:

```yaml
before:
  flow:
    - function: "playwrightInit"
```

```javascript
async function playwrightInit(_context, _event, next) {
  const exec = require('child_process').exec;
  const util = require('util');

  const execPromise = util.promisify(exec)
  const { stdout, stderr } = await execPromise('npx playwright@1.37.1 install');
  console.log('stdout: ', stdout);
  console.log('stderr: ', stderr);
  next();
}
```

**Comment by freyjadomville** [comment]
Unfortunately, this didn't work on Fargate, there's no stdout or stderr to check (unlike when I'm running things locally with this step, which _does_ work), and I therefore believe the npm-cli packages aren't in the image that runs Artillery. A shame, but as a result we're all blocked on this until the `platform-fargate` package can initialise the container with playwright browsers included as a fresh image.

**Comment by tuukkasp** [comment]
I'm getting the same issue on Fargate using 

```
"artillery": "^2.0.0-38",
"@playwright/test": "^1.39.0",
```

**Comment by freyjadomville** [comment]
OK, I've been able to replicate the issue inside a Distrobox of Debian 12 (created via `distrobox create --image debian:stable-backports -n Debian`) and on Arch Linux locally when I don't have a copy of the playwright browsers installed. 

In short, regardless of the browser choice or playwright version, `npx playwright@1.37.1 install` will always install the latest version of the browsers, and for some reason Artillery is always depending on `/ms-playwright/chromium-1076/chrome-linux/chrome`. 

This would explain why we didn't see it before locally and why things are failing during image construction.

Pinging @hassy again for info and good measure - you need to fix this on your side, as we can't vendor this type of thing as a workaround in a library - it's dependent on the obfuscated parts of your artillery-pro offering. Artillery as an organisation should be keeping on top of this type of thing and making sure these types of things work reliably, even for new installations.

**Comment by hassy** [comment]
Going to add more context here for anyone running into similar issue and trying to debug. (Information is accurate at the time of writing and will probably go out of date soon)

Our Playwright integration is currently pinned to v1.39.0 of Playwright.

If you're running a test on Fargate (with [`run-fargate`](https://www.artillery.io/docs/reference/cli/run-fargate)) with the most recent version of Artillery, the Docker image that will be used to execute the tests already includes Playwright v1.39.0 and an installation of Chromium. It should all just work without you needing to do anything. If it doesn't, then it's a bug we need to fix.

If you're running tests elsewhere (e.g. locally, or in GitHub Actions), then you have to make sure to install the browsers. The path where the browsers are expected to be found (e.g. `/ms-playwright/chromium-1076/chrome-linux/chrome`) is set by Playwright itself and not by Artillery, and it changes between version of Playwright, so you have to make sure you're installing v1.39.0 exactly.

**Comment by TestItEver** [comment]
> If you're running a test on Fargate (with [`run-fargate`](https://www.artillery.io/docs/reference/cli/run-fargate)) with the most recent version of Artillery, the Docker image that will be used to execute the tests already includes Playwright v1.39.0 and an installation of Chromium. It should all just work without you needing to do anything. If it doesn't, then it's a bug we need to fix.
> 

Hi @hassy! That is exactly, what we are doing: running a test on Farget with npx artillery run-fargate and get the error " Executable doesn't exist at /ms-playwright/chromium-1076/chrome-linux/chrome". The artillery version is 2.0.0.-37. I will test it with the most recent version.

**Comment by lazzarello** [comment]
> Going to add more context here for anyone running into similar issue and trying to debug. (Information is accurate at the time of writing and will probably go out of date soon)
> 
> Our Playwright integration is currently pinned to v1.39.0 of Playwright. (We'd like to offer the ability to pick any version of Playwright, but as the Playwright team do a good job of maintaining backwards-compatibility, it's not a priority right now.)
> 
> If you're running a test on Fargate (with [`run-fargate`](https://www.artillery.io/docs/reference/cli/run-fargate)) with the most recent version of Artillery, the Docker image that will be used to execute the tests already includes Playwright v1.39.0 and an installation of Chromium. It should all just work without you needing to do anything. If it doesn't, then it's a bug we need to fix.
> 
> If you're running tests elsewhere (e.g. locally, or in GitHub Actions), then you have to make sure to install the browsers. The path where the browsers are expected to be found (e.g. `/ms-playwright/chromium-1076/chrome-linux/chrome`) is set by Playwright itself and not by Artillery, and it changes between version of Playwright, so you have to make sure you're installing v1.39.0 exactly.

This appears to have changed again. I'm new to Playwright and Artillery so it's my first encounter with the issue. It seems like many have worked around it. I'll spend some time trying to do the same.

```
lee@suzanne:~/src/artillery_tutorial$ npm install playwright@v1.39.0

added 2 packages in 4s
lee@suzanne:~/src/artillery_tutorial$ artillery run playwright.yml 
Test run id: tth9d_6eebc96bz85p8kfrghmfjk78m8afg_rqyd
Phase started: unnamed (index: 0, duration: 1s) 15:33:03(-0800)

Phase completed: unnamed (index: 0, duration: 1s) 15:33:04(-0800)

worker error, id: 1 browserType.launch: Executable doesn't exist at /home/lee/.cache/ms-playwright/chromium-1084/chrome-linux/chrome
╔═════════════════════════════════════════════════════════════════════════╗
║ Looks like Playwright Test or Playwright was just installed or updated. ║
║ Please run the following command to download new browsers:              ║
║                                                                         ║
║     npx playwright install                                              ║
║                                                                         ║
║ <3 Playwright Team                                                      ║
╚═════════════════════════════════════════════════════════════════════════╝
    at Array.scenario (/usr/lib/node_modules/artillery/node_modules/artillery-engine-playwright/index.js:67:38)
    at runScenario (/usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/lib/runner.js:373:32)
    at EventEmitter.<anonymous> (/usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/lib/runner.js:203:25)
    at EventEmitter.emit (/usr/lib/node_modules/artillery/node_modules/eventemitter3/index.js:181:35)
    at UniformProcess.<anonymous> (/usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/lib/phases.js:220:12)
    at /usr/lib/node_modules/artillery/node_modules/arrivals/lib/index.js:47:10
    at NanoTimer.setInterval (/usr/lib/node_modules/artillery/node_modules/nanotimer/lib/nanotimer.js:220:5)
    at Immediate._onImmediate (/usr/lib/node_modules/artillery/node_modules/nanotimer/lib/nanotimer.js:203:68) {
  name: 'Error'
}
lee@suzanne:~/src/artillery_tutorial$ ls -ld /home/lee/.cache/ms-playwright/chromium-*
drwxrwxr-x 3 lee lee 4096 Dec 11 11:42 /home/lee/.cache/ms-playwright/chromium-1091
lee@suzanne:~/src/artillery_tutorial$ npx playwright --version
Version 1.39.0
```

**Comment by hassy** [comment]
@lazzarello glad it's working now!

**Comment by BhavnaSoni1991** [comment]
Hi @hassy! I am still getting error "worker error, id: 1 browserType.launch: Executable doesn't exist at /root/.cache/ms-playwright/chromium-1084/chrome-linux/chrome" in pipeline even when I am installing playwright version 1.39.0. 
Below is my Dockerfile.

FROM artilleryio/artillery:2.0.4

RUN npm install -g artillery@2.0.4 artillery-engine-playwright@1.0.0 playwright@1.39.0 && \
        npm cache clean --force && \
        rm -rf /root/.cache && \
        rm -rf /ms-playwright/firefox* && \
        rm -rf /ms-playwright/webkit*

ENTRYPOINT ["/home/node/artillery/bin/run"]
